imports/client/components/BookmarkButton.tsx

```diff
@@ -0,0 +1,50 @@
+import { faStar as faStarRegular } from '@fortawesome/free-regular-svg-icons/faStar';
+import { faStar as faStarSolid } from '@fortawesome/free-solid-svg-icons/faStar';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import React, { useCallback, useState } from 'react';
+import { Button } from 'react-bootstrap';
+import bookmarkPuzzle from '../../methods/bookmarkPuzzle';
+
+interface BookmarkButtonProps<As extends React.ElementType = React.ElementType>
+  extends React.HTMLAttributes<HTMLElement> {
+  puzzleId: string;
+  bookmarked: boolean;
+  as?: As;
+}
+
+type BookmarkButtonType = <
+  As extends React.ElementType = typeof Button
+>(
+  props: BookmarkButtonProps<As> & Omit<React.ComponentPropsWithRef<As>, keyof BookmarkButtonProps>,
+  context?: any,
+) => React.ReactNode;
+
+const BookmarkButton: BookmarkButtonType = React.forwardRef<HTMLElement, BookmarkButtonProps>(({
```

Comment:
For the record, these type shenanigans are pretty heavily inspired by some similar type shenanigans in react-bootstrap.